src/watch.rs

```diff
@@ -51,14 +53,41 @@ impl WatchPathBuf {
 /// react to changes when they occur.
 pub struct Watch {
     /// Event receiver. Process using `Watch::process`.
-    pub rx: chan::Receiver<notify::Result<notify::Event>>,
-    /// OS-based notification when any file we watched changes.
-    notify: RecommendedWatcher,
-    /// The list of files we are watching;
+    pub event_rx: chan::Receiver<Vec<PathBuf>>,
+    watch_tx: chan::Sender<Vec<WatchPathBuf>>,
+}
+
+impl Watch {
+    /// Instantiate a new Watch.
+    pub fn new(logger: slog::Logger) -> Result<Watch, notify::Error> {
+        let (ntfy_tx, notify_rx) = chan::unbounded();
+        let (event_tx, event_rx) = chan::unbounded();
+        let (watch_tx, watch_rx) = chan::unbounded();
+
+        let filter = Filter {
+            notify: Watcher::new(ntfy_tx, Config::default())?,
+            notify_rx,
+            watch_rx,
+            event_tx,
+            watches: HashSet::new(),
+            logger: logger.clone(),
+        };
+
+        thread::spawn(move || filter.run());
```

Comment:
We shouldn’t just spawn threads and ignore their thread handles

Reply:
That was my first instinct but we don't have anything to wait on the thread for (all its results are streamed out), so there's little value to `join`ing.

Once the Watcher gets dropped, its side of the channels are dropped, and the Filter will get those errors (e.g. RecvError) and exit its loop.

In the end, I couldn't find a sensible way to make the captured thread handle not dead code, which is why I ignored it.

Reply:
Check out the code, I added `Async::run_with_stop_signal` to kill the event loop on drop.